python/taichi/_main.py

```diff
@@ -188,31 +194,30 @@ def draw_bounding_box(rind, cind):
         def on_mouse_click_callback(example_name):
             examples_dir = TaichiMain._get_examples_dir()
             script = list(examples_dir.rglob(f"{example_name}.py"))[0]
-            with open(script, "r") as f:
-                try:
-                    import rich.console  # pylint: disable=C0415
-                    import rich.syntax  # pylint: disable=C0415
-                    content = rich.syntax.Syntax.from_path(script,
-                                                           line_numbers=True)
-                    console = rich.console.Console()
-                    console.print(content)
-                except ImportError:
-                    content = f.readlines()
-                    print(content)
+            try:
+                import rich.console  # pylint: disable=C0415
+                import rich.syntax  # pylint: disable=C0415
+                content = rich.syntax.Syntax.from_path(script,
+                                                       line_numbers=True)
+                console = rich.console.Console()
+                console.print(content)
+            except ImportError:
+                with open(script, "r") as f:
+                    shutil.copyfileobj(f, sys.stdout)
 
             self._exec_python_file(script)
 
         while gui.running:
             gui.set_image(gallery_image)
-            if gui.get_events(gui.MOTION):
-                mou_x, mou_y = gui.get_cursor_pos()
-                valid, rind, cind = get_tile_from_mouse(mou_x, mou_y)
-                if valid:
-                    draw_bounding_box(rind, cind)
-                    if gui.get_event(ti.GUI.PRESS, ti.GUI.LMB):
-                        gui.close()
-                        index = cind + rind * ncols
-                        on_mouse_click_callback(examples[index])
+            mou_x, mou_y = gui.get_cursor_pos()
+            if valid_mouse_position(mou_x, mou_y):
+                rind, cind = get_tile_from_mouse(mou_x, mou_y)
+                draw_bounding_box(rind, cind)
+                gui.get_event(ti.GUI.PRESS)
+                if gui.is_pressed(ti.GUI.LMB):
+                    gui.close()
+                    index = cind + rind * ncols
```

Comment:
```
  File "/Users/yuanming/repos/taichi/python/taichi/_main.py", line 219, in gallery
    index = cind + rind * ncols
NameError: name 'ncols' is not defined
```

I get this when my mouse enters the window :-)

Reply:
Oops! Sorry I mistakenly deleted that in a online format editing :cry: :
Added back now.